### PR TITLE
C++: Replace `field -> object` taint rule with `ArrayContent` dataflow

### PR DESCRIPTION
**Pull request by MathiasVP**
This PR uses field flow to implement precise modeling of array writes and reads for IR dataflow.

With these changes to the C++ dataflow library, we can now remove one of the taint rules that causes `field -> object` taint flow (which we want to avoid before we embark on https://github.com/github/codeql-c-analysis-team/issues/103 to avoid field-to-unrelated-field flow).

### CODE REVIEW

**Comment by MathiasVP** [comment]
The [CPP-differences](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1423/) came back surprisingly positive!

There are no result changes (even though we swap many taint flows out with data flows), and it doesn't look my generous definition of `arrayReadStep` leads to any performance regressions! (there's an 8% slowdown on `php` caused by `SignedOverflowCheck.ql`, but that query doesn't use any dataflow, so I'm attributing this to noise).

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:250]: These rules seem overly syntactic, but we can always tweak them later if there's a need. The important thing is they're guaranteed not to overlap with `getWrittenField`.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:304]: This definition seems overly broad. Can we restrict the address operand to particular syntactic forms like in `arrayStoreStepChi` and ensure there is no overlap with `fieldReadStep`?

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:394]: Do I understand it correctly that this class and `PointerStoreNode` have to correspond to the implementation of `arrayStoreStepChi`? If so, we have to make that clear for future maintainers. Either they can share code or there can at least be a comment in all three places referring to each other.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:253]: Check `isResultConflated` here?

File: cpp/ql/test/library-tests/dataflow/fields/ir-path-flow.expected
Comment (jbj) [inline cpp/ql/test/library-tests/dataflow/fields/ir-path-flow.expected:93]: Strings like `Chi` and `Store` are user-visible, so we'll have to give them better `toString`s at some point. But I can see that the problematic strings were already present before this PR, so that can be follow-up work.

**Review by jbj — COMMENTED** [review]
Overall I'm very excited about this PR. We'll have to figure out how it fits into the collection taint modelling we're doing these days.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:250]: I agree. The `PointerAdd` case I think we can live with, but requiring the presence of a `CopyValue` instruction as a direct successor to the `Load` instruction is quite fragile. For instance, for this store:
```
int source();
void f(char* s) {
	*(int*)s = source();
}
```
there's a `Convert` instruction in between the `Load` and the `CopyValue`, which means that this is not recognized as a store step by `arrayStoreStepChi`.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:304]: This was also my main worry when I was CPP-difference, but luckily didn't appear to be a problem. For explicit array and pointer reads (i.e., `x[i]`) we could restrict it by requiring that the source address instruction is a `PointerAdd` instruction. But for an example like:
```cpp
void f(int* s) {
  *s = source();
}

void test() {
  int x;
  f(&x);
  sink(x);
}
```
(where the read step is currently the load of `x` in `sink(x)`) I don't see how to syntactically restrict the predicate.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:398]: Yes, this is correct. I've added a couple of comments in 50ad4cfec4e80387fa67e8f9a21eeb45128a20fd to make this correspondence explicit.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:257]: It will always be a non-conflated `Chi` because the char preds for `ArrayStoreNode` and `PointerStoreNode` require the instruction to be non-conflated. Performance on `linux` seems to be the same, but if you think it's important for readability I can add the check. I've added a comment saying that it'll always be non-conflated in 50ad4cfec4e80387fa67e8f9a21eeb45128a20fd.

Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:257]: I'm okay with the readability as long as the performance is the same.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll
Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:308]: But can you add something like `and not fieldReadStep(node1, _, _)` to ensure there is no overlap?

Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:308]: Hm, I was hoping your solution would Just Work, but the following shows that we need to do something else, I think:
```cpp
void taint_a_ptr(int* px) {
  *px = source();
}

void test_outer_with_ptr() {
  Point p;
  f(&p.x);
  sink(p.x);
}
```

Here, the load of `p.x` in `sink(p.x)` actually looks like a `fieldReadStep`, but (as `*px = source()` matches `arrayStoreStepChi`), we actually need an `arrayReadStep` in this case.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll:271]: I think you can remove the same rule from IR `TaintTrackingUtil.qll`.

Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll:271]: Good point. I've removed it in 0b97a4a182bc19a9d32cea6364cc3e051be439a0.

**Comment by MathiasVP** [comment]
I didn't notice until now that there's actually a result change in our tests for `UncontrolledFormatStringl.ql` in `argvLocal.c`, where we no longer report these two results after 78b24b76a09d680ac6581597f0ecfe25596ed18f (which removed `array element -> array` taint):

```cpp
// BAD: i5 value comes from argv
char i5[5012];
i5[0] = argv[1][0];
printf(i5);
printWrapper(i5);
	
// BAD: i5 value comes from argv
printf(i5 + 1);
printWrapper(i5 + 1);
```

I now think that we should:
- Undo that particular commit, and only remove the `t instanceof UnknownType` case from `DefaultTaintTracking`
- Undo 0b97a4a182bc19a9d32cea6364cc3e051be439a0 which removed the same rule from `TaintTrackingUtil` that was removed in )
The good thing is that now those two rules will be completely identical. Once we have the `field-to-object-flow-at-sink` flow as mentioned in https://github.com/github/codeql-c-analysis-team/issues/103 we can remove the rule completely.

**Comment by jbj** [comment]
> Which are legitimate uses of field -> object flow.

How so? I don't see any fields there, so I'd classify it as element -> array.

> How does this sound?

SGTM.

**Comment by MathiasVP** [comment]
> How so? I don't see any fields there, so I'd classify it as element -> array.

Right. I meant `element -> array`. catch!

**Comment by MathiasVP** [comment]
This PR should be good to go now! The remaining test failures are accepted in the internal repo (which is green).

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll
Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:308]: We discussed this and decided it doesn't need to block this PR.

**Review by jbj — APPROVED** [review]

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:308]: As reported in https://github.com/github/codeql/pull/4298 this is a source of field conflation (which I would've discovered if I had tried to replace `p.x` with `p.y` in the example above 😢). This is fixed by https://github.com/github/codeql/pull/4302.

**Pull request merged**
